mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,804 @@
+// The pass is triggered by the command line flag -affine-loop-interchange.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+#include "mlir/Analysis/AffineAnalysis.h"
+#include "mlir/Analysis/Utils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/Passes.h"
+#include "mlir/Transforms/LoopUtils.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include <algorithm>
+#include <cmath>
+
+using namespace mlir;
+namespace {
+struct LoopInterchange : public AffineLoopInterchangeBase<LoopInterchange> {
+  void runOnFunction() override;
+  void handleImperfectlyNestedAffineLoops(Operation &funcOp);
+};
+}
+
+/// Returns True if any affine.if op found in the loop nest rooted at `forOp`
+static bool hasAffineIfStatement(AffineForOp &forOp) {
+  auto walkResult =
+      forOp.walk([&](AffineIfOp op) { return WalkResult::interrupt(); });
+  return walkResult.wasInterrupted();
+}
+
+/// Checks if this `loopNest` has a rectangular shaped iteration space.
+static bool isRectangularAffineForLoopNest(ArrayRef<AffineForOp> loopNest) {
+  for (AffineForOp forOp : loopNest) {
+    if (!forOp.hasConstantUpperBound() || !forOp.hasConstantLowerBound())
+      return false;
+  }
+  return true;
+}
+
+/// Returns true if this entire column of `matrix` is zero.
+static bool checkColumnIsZero(ArrayRef<SmallVector<int64_t, 4>> matrix,
+                              unsigned column) {
+  for (const SmallVector<int64_t, 4> &row : matrix) {
+    if (row[column] != 0)
+      return false;
+  }
+  return true;
+}
+
+/// Fills `row` with the coefficients of loopIVs in `expr`. Any constant terms
+/// encountered in `expr` are added to `constantVectorValue`. Every value in
+/// `operands` should be a loopIV or a terminal symbol.
+static void prepareCoeffientRow(AffineExpr &expr, SmallVector<int64_t, 4> &row,
+                                int64_t &constantVectorValue,
+                                ArrayRef<Value> operands,
+                                DenseMap<Value, unsigned> &loopIndexMap) {
+  // TODO: Implement support for terminal symbols in `expr`.
+  switch (expr.getKind()) {
+  case AffineExprKind::Add: {
+    // Is this sub-expr a constant? If yes, no need to modify `row`. Start by
+    // assuming the sub-expr is not a constant.
+    bool isConstSubExpr = false;
+    AffineBinaryOpExpr addExpr = expr.cast<AffineBinaryOpExpr>();
+    AffineExpr lhs = addExpr.getLHS();
+    AffineExpr rhs = addExpr.getRHS();
+    unsigned lhsPosition = 0;
+    unsigned rhsPosition = 0;
+    
+    switch (lhs.getKind()) {
+    case AffineExprKind::DimId: {
+      auto dim = lhs.cast<AffineDimExpr>();
+      lhsPosition = loopIndexMap[operands[dim.getPosition()]];
+      break;
+    }
+    case AffineExprKind::Constant: {
+      constantVectorValue += rhs.cast<AffineConstantExpr>().getValue();
+      isConstSubExpr = true;
+    }
```

Comment:
Missing remaining cases - use `if` if it's just these two.

Reply:
Fixed. Thanks.